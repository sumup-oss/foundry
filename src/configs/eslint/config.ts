/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import process from 'process';

import { flow, mergeWith, isArray, isObject, isEmpty, uniq } from 'lodash/fp';

import { Options, Language, Environment, Framework } from '../../types/shared';

type EslintOptions = Pick<
  Options,
  'language' | 'environments' | 'frameworks' | 'openSource'
>;
// NOTE: Using the Linter.Config interface from Eslint causes errors
//       and I couldn't figure out how to fix them. â€” @connor_baer
type EslintConfig = any;

export const customizeConfig = mergeWith(customizer);

function customizer(objValue: any, srcValue: any, key: string): any {
  if (isArray(objValue)) {
    return uniq([...objValue, ...srcValue]);
  }
  if (isObject(objValue)) {
    return key === 'rules' ? { ...objValue, ...srcValue } : undefined;
  }
  return undefined;
}

const base = {
  root: true,
  extends: ['eslint:recommended', 'plugin:prettier/recommended'],
  plugins: ['prettier'],
  rules: {
    'curly': ['error', 'all'],
    'no-use-before-define': 'off',
    'no-confusing-arrow': 'off',
    'max-len': [
      'error',
      {
        code: 80,
        tabWidth: 2,
        ignoreComments: true,
        ignoreUrls: true,
        ignoreStrings: true,
        ignoreTemplateLiterals: true,
        ignoreRegExpLiterals: true,
        ignorePattern:
          '^(?:import\\s|export\\s|\\s*it(?:\\.(?:skip|only))?\\()',
      },
    ],
    'no-underscore-dangle': [
      'error',
      { allow: ['__DEV__', '__PRODUCTION__', '__TEST__'] },
    ],
    'import/prefer-default-export': 'off',
    'import/order': ['error', { 'newlines-between': 'always' }],
    // The rules below are already covered by prettier.
    'quote-props': 'off',
    'comma-dangle': 'off',
    'object-curly-newline': 'off',
    'implicit-arrow-linebreak': 'off',
    'function-paren-newline': 'off',
    'operator-linebreak': 'off',
    'indent': 'off',
  },
  globals: {
    __DEV__: true,
    __PRODUCTION__: true,
    __TEST__: true,
  },
  overrides: [
    {
      files: [
        '**/*.story.*',
        '**/*.stories.*',
        '**/setupTests.*',
        '**/test-utils.*',
      ],
      rules: {
        'import/no-extraneous-dependencies': 'off',
      },
    },
  ],
};

function customizeLanguage(language?: Language): EslintConfig {
  const languageMap = {
    [Language.TYPESCRIPT]: {
      extends: [
        'airbnb-typescript/base',
        'plugin:@typescript-eslint/eslint-recommended',
        'plugin:@typescript-eslint/recommended',
        'plugin:@typescript-eslint/recommended-requiring-type-checking',
        'prettier/@typescript-eslint',
      ],
      plugins: ['@typescript-eslint'],
      parser: '@typescript-eslint/parser',
      parserOptions: {
        tsconfigRootDir: process.cwd(),
        project: ['./tsconfig.json'],
        sourceType: 'module',
        ecmaVersion: 6,
        ecmaFeatures: {
          modules: true,
        },
      },
      settings: {
        'import/resolver': {
          node: {
            extensions: ['.js', '.jsx', '.ts', '.tsx'],
          },
        },
      },
      rules: {
        '@typescript-eslint/no-use-before-define': [
          'error',
          { functions: false },
        ],
        'react/prop-types': 'off',
      },
      overrides: [
        {
          files: ['**/*.js'],
          rules: {
            '@typescript-eslint/explicit-function-return-type': 'off',
          },
        },
        {
          files: ['**/*.d.ts'],
          rules: {
            'spaced-comment': 'off',
            'node/no-extraneous-import': 'off',
            'import/no-extraneous-dependencies': [
              'error',
              { devDependencies: true },
            ],
          },
        },
        {
          files: ['**/*.spec.*'],
          rules: {
            '@typescript-eslint/no-var-requires': 'off',
          },
        },
      ],
    },
    [Language.JAVASCRIPT]: {
      extends: ['airbnb-base'],
      parser: 'babel-eslint',
      parserOptions: {
        sourceType: 'module',
        ecmaVersion: 6,
        ecmaFeatures: {
          modules: true,
          impliedStrict: true,
        },
        allowImportExportEverywhere: true,
      },
    },
  };
  return (config: EslintConfig): EslintConfig => {
    if (!language) {
      return config;
    }
    const overrides = languageMap[language];
    return customizeConfig(config, overrides);
  };
}

function customizeEnv(environments?: Environment[]): EslintConfig {
  const environmentMap = {
    [Environment.BROWSER]: {
      env: { browser: true },
    },
    [Environment.NODE]: {
      extends: ['plugin:node/recommended'],
      env: { node: true },
      rules: {
        // We don't know if the user's source code is using EJS or CJS.
        'node/no-unsupported-features/es-syntax': 'off',
        // This rule breaks when used in combination with TypeScript
        // and is already covered by similar Eslint rules.
        'node/no-missing-import': 'off',
        // This rule is already covered by similar Eslint rules.
        'node/no-extraneous-import': 'off',
      },
      overrides: [
        {
          files: ['**/*.spec.*', '**/setupTests.*', '**/test-utils.*'],
          rules: {
            'node/no-unpublished-import': 'off',
          },
        },
      ],
    },
  };
  return (config: EslintConfig): EslintConfig => {
    if (!environments || isEmpty(environments)) {
      return config;
    }
    return environments.reduce((acc, environment: Environment) => {
      const overrides = environmentMap[environment];
      return customizeConfig(acc, overrides);
    }, config);
  };
}

function customizeFramework(frameworks?: Framework[]): EslintConfig {
  const frameworkMap = {
    [Framework.REACT]: {
      extends: [
        'plugin:react/recommended',
        'plugin:jsx-a11y/recommended',
        'prettier/react',
      ],
      plugins: ['react', 'react-hooks', 'jsx-a11y'],
      rules: {
        'react-hooks/rules-of-hooks': 'error',
        'react-hooks/exhaustive-deps': 'warn',
      },
      parserOptions: { ecmaFeatures: { jsx: true } },
      settings: { react: { version: 'detect' } },
    },
    [Framework.EMOTION]: {
      plugins: ['emotion'],
      rules: {
        'emotion/jsx-import': 'off',
        'emotion/no-vanilla': 'error',
        'emotion/import-from-emotion': 'error',
        'emotion/styled-import': 'error',
      },
    },
    [Framework.JEST]: {
      extends: ['plugin:jest/recommended'],
      plugins: ['jest'],
      overrides: [
        {
          files: ['**/*spec.*'],
          globals: {
            render: true,
            create: true,
            renderToHtml: true,
            fireEvent: true,
            userEvent: true,
            wait: true,
            act: true,
            actHook: true,
            renderHook: true,
            axe: true,
          },
          env: { 'jest/globals': true },
        },
      ],
    },
    [Framework.CYPRESS]: {
      extends: ['plugin:cypress/recommended'],
      plugins: ['cypress'],
      overrides: [
        {
          files: ['**/*spec.*', 'e2e/**/*'],
          env: { 'cypress/globals': true },
        },
      ],
    },
  };
  return (config: EslintConfig): EslintConfig => {
    if (!frameworks || isEmpty(frameworks)) {
      return config;
    }
    return frameworks.reduce((acc, framework: Framework) => {
      const overrides = frameworkMap[framework];
      return customizeConfig(acc, overrides);
    }, config);
  };
}

function addCopyrightNotice(openSource?: boolean): EslintConfig {
  return (config: EslintConfig): EslintConfig => {
    if (!openSource) {
      return config;
    }
    const copyrightNotice = {
      plugins: ['notice'],
      rules: {
        'notice/notice': [
          'error',
          {
            template: `/**
 * Copyright <%= YEAR %>, <%= NAME %>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

`,
            templateVars: { NAME: 'SumUp Ltd.' },
            varRegexps: { NAME: /SumUp Ltd\./ },
            onNonMatchingHeader: 'prepend',
          },
        ],
      },
    };
    return customizeConfig(config, copyrightNotice);
  };
}

function applyOverrides(overrides: EslintConfig): EslintConfig {
  return (config: EslintConfig): EslintConfig =>
    customizeConfig(config, overrides);
}

export function createConfig(
  options: EslintOptions = {},
  overrides: EslintConfig = {},
): EslintConfig {
  return flow(
    customizeLanguage(options.language),
    customizeEnv(options.environments),
    customizeFramework(options.frameworks),
    addCopyrightNotice(options.openSource),
    applyOverrides(overrides),
  )(base);
}
