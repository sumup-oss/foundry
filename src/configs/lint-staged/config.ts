/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Language, Options } from '../../types/shared';

type LintStagedOptions = Pick<Options, 'language'>;
type LinterCommand = string | string[];
type LinterFn = (filenames: string[]) => LinterCommand;

interface LintStagedConfig {
  [key: string]: LinterCommand | LinterFn;
}

export const javascript: LintStagedConfig = {
  '*.(js|jsx|json)': ['foundry run eslint --fix'],
};

export const typescript: LintStagedConfig = {
  '*.(js|jsx|json|ts|tsx)': ['foundry run eslint --fix'],
  '*.(ts|tsx)': () => 'tsc -p tsconfig.json --noEmit',
};

const LANGUAGES = {
  [Language.JAVASCRIPT]: javascript,
  [Language.TYPESCRIPT]: typescript,
};

export function config(
  options: LintStagedOptions = {},
  overrides: LintStagedConfig = {},
): LintStagedConfig {
  const { language = Language.JAVASCRIPT } = options;
  const baseConfig = LANGUAGES[language];
  return { ...baseConfig, ...overrides };
}
