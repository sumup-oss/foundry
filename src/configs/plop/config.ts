/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { existsSync } from 'fs';
import { relative, join } from 'path';

import { NodePlopAPI, ActionConfig } from 'plop';

import * as logger from '../../lib/logger';
import { Language } from '../../types/shared';

interface PlopOptions {
  templateDir?: string;
  language: Language;
}

enum ComponentType {
  STYLED = 'styled',
  FUNCTIONAL = 'functional',
  CLASS = 'class',
}

enum FileType {
  COMPONENT = 'component',
  COMPONENT_SPEC = 'component-spec',
  STORY = 'story',
  SERVICE = 'service',
  SERVICE_SPEC = 'service-spec',
  INDEX = 'index',
}

interface ActionOptions {
  name: string;
  componentType: ComponentType;
  files: FileType[];
  destinationPath: string;
}

const ERRORS = {
  INVALID_FILE_TYPE: (file: string): string =>
    `We don't have templates for "${file}" files, ðŸ˜ž.`,
  INVALID_DESTINATION: (path: string): string =>
    `We couldn't find the destination folder ${path} in your project, ðŸ¤·.`,
};

export function config(options: PlopOptions) {
  return (plop: NodePlopAPI): void => {
    plop.setHelper('eq', (a: any, b: any): boolean => a === b);
    plop.setHelper('not', (a: any, b: any): boolean => a !== b);

    const pascalCase = plop.getHelper('pascalCase');

    /**
     * Generating React components and helper files.
     */
    plop.setGenerator('component', {
      description: 'React component',
      prompts: [
        // Component name
        {
          type: 'input',
          name: 'name',
          message: "What's the name of your component?",
        },
        // Component type
        {
          type: 'list',
          name: 'componentType',
          message: 'What type of component do you need?',
          choices: [
            {
              name: 'Styled',
              value: ComponentType.STYLED,
            },
            {
              name: 'Functional',
              value: ComponentType.FUNCTIONAL,
            },
            {
              name: 'Class',
              value: ComponentType.CLASS,
            },
          ],
          default: 'functional',
        },
        // Get the path
        {
          type: 'input',
          name: 'destinationPath',
          message: 'Where would you like to put your component?',
          default: relative(
            process.cwd(),
            `${plop.getPlopfilePath()}/src/components`,
          ),
        },
        // Files
        {
          type: 'checkbox',
          name: 'files',
          message: 'Which files does your component need?',
          choices: [
            {
              name: 'Component',
              value: FileType.COMPONENT,
              checked: true,
            },
            {
              name: 'Component spec',
              value: FileType.COMPONENT_SPEC,
              checked: true,
            },
            {
              name: 'Story',
              value: FileType.STORY,
              checked: false,
            },
            {
              name: 'Service',
              value: FileType.SERVICE,
              checked: false,
            },
            {
              name: 'Service spec',
              value: FileType.SERVICE_SPEC,
              checked: false,
            },
          ],
        },
      ],
      actions: ({
        name: componentName,
        componentType,
        files,
        destinationPath,
      }: ActionOptions): ActionConfig[] => {
        const plopfilePath = plop.getPlopfilePath();
        const absDestPath = join(plopfilePath, destinationPath);

        if (!existsSync(absDestPath)) {
          raiseErrorAndExit(ERRORS.INVALID_DESTINATION(absDestPath));
        }

        const capitalizedName = pascalCase(componentName);
        const allFiles = files.includes(FileType.COMPONENT)
          ? files.concat(FileType.INDEX)
          : files;

        return allFiles.reduce((acc, file) => {
          if (!Object.values(FileType).includes(file)) {
            raiseErrorAndExit(ERRORS.INVALID_FILE_TYPE(file));
          }

          const templateFileName =
            file === FileType.COMPONENT
              ? getComponentTemplateName(componentType)
              : `${file}.hbs`;

          return [
            ...acc,
            {
              type: 'add',
              path: join(
                absDestPath,
                capitalizedName,
                getFileName(options.language, file, componentName),
              ),
              templateFile: getTemplatePath(
                options.language,
                options.templateDir,
                plopfilePath,
                templateFileName,
              ),
            },
          ];
        }, [] as ActionConfig[]);
      },
    });
  };
}

function raiseErrorAndExit(message: string): never {
  logger.error([message, 'Please try again. ðŸ‘‹'].join(' '));
  process.exit(1);
}

function getComponentTemplateName(type: ComponentType): string {
  const templateNameMap = {
    [ComponentType.STYLED]: 'styled-component.hbs',
    [ComponentType.FUNCTIONAL]: 'functional-component.hbs',
    [ComponentType.CLASS]: 'class-component.hbs',
  };
  return templateNameMap[type];
}

const JS_EXTENSIONS = {
  [Language.JAVASCRIPT]: 'js',
  [Language.TYPESCRIPT]: 'ts',
};

const JSX_EXTENSIONS = {
  [Language.JAVASCRIPT]: 'js',
  [Language.TYPESCRIPT]: 'tsx',
};

function getFileName(
  language: Language,
  fileType: FileType,
  componentName: string,
): string {
  const jsExtension = JS_EXTENSIONS[language];
  const jsxExtension = JSX_EXTENSIONS[language];
  const fileNameMap = {
    [FileType.COMPONENT]: `${componentName}.${jsxExtension}`,
    [FileType.COMPONENT_SPEC]: `${componentName}.spec.${jsxExtension}`,
    [FileType.STORY]: `${componentName}.story.${jsxExtension}`,
    [FileType.SERVICE]: `${componentName}Service.${jsExtension}`,
    [FileType.SERVICE_SPEC]: `${componentName}Service.spec.${jsExtension}`,
    [FileType.INDEX]: `index.${jsExtension}`,
  };
  return fileNameMap[fileType];
}

const LANGUAGE_DIRS = {
  [Language.JAVASCRIPT]: 'js',
  [Language.TYPESCRIPT]: 'ts',
};

function getTemplatePath(
  language: Language,
  templateDir: string | undefined,
  plopDir: string,
  templateFileName: string,
): string {
  if (templateDir) {
    const customPath = join(plopDir, templateDir, templateFileName);
    if (existsSync(customPath)) {
      return customPath;
    }
  }

  const languageDir = LANGUAGE_DIRS[language];

  return join(__dirname, 'templates', languageDir, templateFileName);
}
