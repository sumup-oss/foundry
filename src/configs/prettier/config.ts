/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Options as PrettierConfig } from 'prettier';

export interface PrettierOptions {}

const base: PrettierConfig = {
  printWidth: 80,
  tabWidth: 2,
  useTabs: false,
  semi: true,
  singleQuote: true,
  quoteProps: 'consistent',
  trailingComma: 'none',
  jsxSingleQuote: false,
  bracketSpacing: true,
  jsxBracketSameLine: false,
  arrowParens: 'always',
  endOfLine: 'lf'
};

export function config(
  options: PrettierOptions = {},
  overrides: PrettierConfig = {}
) {
  return { ...base, ...overrides };
}
