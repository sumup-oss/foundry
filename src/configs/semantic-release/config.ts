/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Options } from '../../types/shared';

type SemanticReleaseOptions = Pick<Options, 'publish'>;

type Branch = string | { name: string; prerelease: boolean };

interface SemanticReleaseConfig {
  extends?: string | string[];
  branches?: Branch[];
  plugins?: string[];
  tagFormat?: string;
  repositoryUrl?: string;
  dryRun?: boolean;
  ci?: boolean;
}

const base: SemanticReleaseConfig = {
  branches: [
    '+([0-9])?(.{+([0-9]),x}).x',
    'main',
    'next',
    { name: 'alpha', prerelease: true },
    { name: 'beta', prerelease: true },
    { name: 'canary', prerelease: true },
    // @deprecated, remove in next major release
    'master',
  ],
  plugins: [
    '@semantic-release/commit-analyzer',
    '@semantic-release/release-notes-generator',
    '@semantic-release/github',
  ],
};

const npm: SemanticReleaseConfig = {
  ...base,
  plugins: [
    '@semantic-release/commit-analyzer',
    '@semantic-release/release-notes-generator',
    '@semantic-release/npm',
    '@semantic-release/github',
  ],
};

export function config(
  options: SemanticReleaseOptions = {},
  overrides: SemanticReleaseConfig = {},
): SemanticReleaseConfig {
  const { publish = false } = options;
  const baseConfig = publish ? npm : base;
  return { ...baseConfig, ...overrides };
}
