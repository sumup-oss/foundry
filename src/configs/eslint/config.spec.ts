/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Target } from '../../types/shared';
import { getAllChoiceCombinations } from '../../lib/choices';

import { overwritePresets, config } from './config';

describe('eslint', () => {
  describe('overwritePresets', () => {
    it('should merge two configurations overwriting existing presets', () => {
      const config = {
        extends: ['some-preset'],
        rules: {
          'some-custom-rule': false
        }
      };
      const newConfig = {
        extends: ['other-preset']
      };
      const actual = overwritePresets(config, newConfig);
      expect(actual).toEqual(
        expect.objectContaining({
          extends: ['other-preset'],
          rules: {
            'some-custom-rule': false
          }
        })
      );
    });

    it('should overwrite the extends of the base config with the custom config', () => {
      const base = {
        extends: [
          'airbnb-base',
          'plugin:jest/recommended',
          'plugin:prettier/recommended'
        ]
      };
      const custom = {
        extends: [
          'airbnb-base',
          'plugin:react/recommended',
          'plugin:jsx-a11y/recommended',
          'prettier/react'
        ]
      };
      const expected = custom;
      const actual = overwritePresets(base, custom);
      expect(actual).toEqual(expected);
    });

    it('should merge the rules of the base config with the custom config', () => {
      const base = {
        rules: {
          'no-use-before-define': ['error', { functions: false }],
          'curly': ['error', 'all'],
          'no-underscore-dangle': [
            'error',
            { allow: ['__DEV__', '__PRODUCTION__'] }
          ]
        }
      };
      const custom = {
        rules: {
          'react-hooks/rules-of-hooks': 'error',
          'react-hooks/exhaustive-deps': 'warn',
          'no-underscore-dangle': ['warning', { allow: ['__resourcePath'] }]
        }
      };
      const expected = {
        rules: {
          'no-use-before-define': ['error', { functions: false }],
          'curly': ['error', 'all'],
          'no-underscore-dangle': ['warning', { allow: ['__resourcePath'] }],
          'react-hooks/rules-of-hooks': 'error',
          'react-hooks/exhaustive-deps': 'warn'
        }
      };
      const actual = overwritePresets(base, custom);
      expect(actual).toEqual(expected);
    });
  });

  describe('with options', () => {
    const matrix = getAllChoiceCombinations({ target: Target });

    it.each(matrix)('should return a config for %o', (options) => {
      const actual = config(options);
      expect(actual).toMatchSnapshot();
    });
  });

  describe('with overrides', () => {
    it('should override the default config', () => {
      const options = undefined;
      const overrides = {
        extends: ['prettier/react']
      };
      const actual = config(options, overrides);
      expect(actual).toEqual(
        expect.objectContaining({
          extends: ['prettier/react']
        })
      );
    });
  });
});
