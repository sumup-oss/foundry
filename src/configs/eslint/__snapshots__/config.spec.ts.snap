// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eslint with options should return a config for { target: 'Browser' } 1`] = `
Object {
  "extends": Array [
    "airbnb-base",
    "plugin:jest/recommended",
    "plugin:prettier/recommended",
  ],
  "globals": Object {
    "__DEV__": true,
    "__PRODUCTION__": true,
    "__TEST__": true,
  },
  "overrides": Array [
    Object {
      "env": Object {
        "cypress/globals": true,
        "jest/globals": true,
      },
      "files": Array [
        "**/*spec.js",
        "e2e/**/*.js",
      ],
      "globals": Object {
        "axe": true,
        "create": true,
        "mount": true,
        "render": true,
        "renderToHtml": true,
        "shallow": true,
      },
      "rules": Object {
        "max-len": Array [
          "error",
          Object {
            "code": 80,
            "ignoreComments": true,
            "ignorePattern": "^\\\\s*it(?:\\\\.(?:skip|only))?\\\\(",
            "ignoreUrls": true,
            "tabWidth": 2,
          },
        ],
      },
    },
    Object {
      "files": Array [
        "*.config.js",
        ".*rc.js",
      ],
      "rules": Object {
        "import/no-extraneous-dependencies": Array [
          "error",
          Object {
            "devDependencies": true,
          },
        ],
      },
    },
  ],
  "parser": "babel-eslint",
  "parserOptions": Object {
    "allowImportExportEverywhere": true,
    "ecmaFeatures": Object {
      "ecmaVersion": 2017,
      "impliedStrict": true,
    },
  },
  "plugins": Array [
    "prettier",
    "jest",
    "cypress",
    "notice",
  ],
  "rules": Object {
    "curly": Array [
      "error",
      "all",
    ],
    "max-len": Array [
      "error",
      Object {
        "code": 80,
        "ignoreComments": true,
        "ignoreUrls": true,
        "tabWidth": 2,
      },
    ],
    "no-underscore-dangle": Array [
      "error",
      Object {
        "allow": Array [
          "__DEV__",
          "__PRODUCTION__",
          "__TEST__",
        ],
      },
    ],
    "no-use-before-define": Array [
      "error",
      Object {
        "functions": false,
      },
    ],
    "notice/notice": Array [
      "error",
      Object {
        "onNonMatchingHeader": "report",
        "template": "/**
 * Copyright <%= YEAR %>, <%= NAME %>
 * Licensed under the Apache License, Version 2.0 (the \\"License\\");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an \\"AS IS\\" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

",
        "templateVars": Object {
          "NAME": "SumUp Ltd.",
        },
        "varRegexps": Object {
          "NAME": /SumUp Ltd\\\\\\./,
        },
      },
    ],
  },
}
`;

exports[`eslint with options should return a config for { target: 'Node' } 1`] = `
Object {
  "env": Object {
    "node": true,
  },
  "extends": Array [
    "airbnb-base",
    "plugin:jest/recommended",
    "plugin:prettier/recommended",
  ],
  "globals": Object {
    "__DEV__": true,
    "__PRODUCTION__": true,
    "__TEST__": true,
  },
  "overrides": Array [
    Object {
      "env": Object {
        "cypress/globals": true,
        "jest/globals": true,
      },
      "files": Array [
        "**/*spec.js",
        "e2e/**/*.js",
      ],
      "globals": Object {
        "axe": true,
        "create": true,
        "mount": true,
        "render": true,
        "renderToHtml": true,
        "shallow": true,
      },
      "rules": Object {
        "max-len": Array [
          "error",
          Object {
            "code": 80,
            "ignoreComments": true,
            "ignorePattern": "^\\\\s*it(?:\\\\.(?:skip|only))?\\\\(",
            "ignoreUrls": true,
            "tabWidth": 2,
          },
        ],
      },
    },
    Object {
      "files": Array [
        "*.config.js",
        ".*rc.js",
      ],
      "rules": Object {
        "import/no-extraneous-dependencies": Array [
          "error",
          Object {
            "devDependencies": true,
          },
        ],
      },
    },
  ],
  "parser": "babel-eslint",
  "parserOptions": Object {
    "allowImportExportEverywhere": true,
    "ecmaFeatures": Object {
      "ecmaVersion": 2017,
      "impliedStrict": true,
    },
  },
  "plugins": Array [
    "prettier",
    "jest",
    "cypress",
    "notice",
  ],
  "rules": Object {
    "curly": Array [
      "error",
      "all",
    ],
    "max-len": Array [
      "error",
      Object {
        "code": 80,
        "ignoreComments": true,
        "ignoreUrls": true,
        "tabWidth": 2,
      },
    ],
    "no-underscore-dangle": Array [
      "error",
      Object {
        "allow": Array [
          "__DEV__",
          "__PRODUCTION__",
          "__TEST__",
        ],
      },
    ],
    "no-use-before-define": Array [
      "error",
      Object {
        "functions": false,
      },
    ],
    "notice/notice": Array [
      "error",
      Object {
        "onNonMatchingHeader": "report",
        "template": "/**
 * Copyright <%= YEAR %>, <%= NAME %>
 * Licensed under the Apache License, Version 2.0 (the \\"License\\");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an \\"AS IS\\" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

",
        "templateVars": Object {
          "NAME": "SumUp Ltd.",
        },
        "varRegexps": Object {
          "NAME": /SumUp Ltd\\\\\\./,
        },
      },
    ],
  },
}
`;

exports[`eslint with options should return a config for { target: 'React' } 1`] = `
Object {
  "env": Object {
    "browser": true,
  },
  "extends": Array [
    "airbnb-base",
    "plugin:react/recommended",
    "plugin:jsx-a11y/recommended",
    "plugin:prettier/recommended",
    "prettier/react",
  ],
  "globals": Object {
    "__DEV__": true,
    "__PRODUCTION__": true,
    "__TEST__": true,
  },
  "overrides": Array [
    Object {
      "env": Object {
        "cypress/globals": true,
        "jest/globals": true,
      },
      "files": Array [
        "**/*spec.js",
        "e2e/**/*.js",
      ],
      "globals": Object {
        "axe": true,
        "create": true,
        "mount": true,
        "render": true,
        "renderToHtml": true,
        "shallow": true,
      },
      "rules": Object {
        "max-len": Array [
          "error",
          Object {
            "code": 80,
            "ignoreComments": true,
            "ignorePattern": "^\\\\s*it(?:\\\\.(?:skip|only))?\\\\(",
            "ignoreUrls": true,
            "tabWidth": 2,
          },
        ],
      },
    },
    Object {
      "files": Array [
        "*.config.js",
        ".*rc.js",
      ],
      "rules": Object {
        "import/no-extraneous-dependencies": Array [
          "error",
          Object {
            "devDependencies": true,
          },
        ],
      },
    },
  ],
  "parser": "babel-eslint",
  "parserOptions": Object {
    "allowImportExportEverywhere": true,
    "ecmaFeatures": Object {
      "ecmaVersion": 2017,
      "impliedStrict": true,
      "jsx": true,
    },
  },
  "plugins": Array [
    "prettier",
    "jest",
    "cypress",
    "notice",
    "react",
    "react-hooks",
    "jsx-a11y",
    "emotion",
  ],
  "rules": Object {
    "curly": Array [
      "error",
      "all",
    ],
    "emotion/import-from-emotion": "error",
    "emotion/jsx-import": "error",
    "emotion/no-vanilla": "error",
    "emotion/styled-import": "error",
    "import/no-extraneous-dependencies": Array [
      "error",
      Object {
        "devDependencies": true,
      },
    ],
    "max-len": Array [
      "error",
      Object {
        "code": 80,
        "ignoreComments": true,
        "ignoreUrls": true,
        "tabWidth": 2,
      },
    ],
    "no-underscore-dangle": Array [
      "error",
      Object {
        "allow": Array [
          "__DEV__",
          "__PRODUCTION__",
          "__TEST__",
        ],
      },
    ],
    "no-use-before-define": Array [
      "error",
      Object {
        "functions": false,
      },
    ],
    "notice/notice": Array [
      "error",
      Object {
        "onNonMatchingHeader": "report",
        "template": "/**
 * Copyright <%= YEAR %>, <%= NAME %>
 * Licensed under the Apache License, Version 2.0 (the \\"License\\");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an \\"AS IS\\" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

",
        "templateVars": Object {
          "NAME": "SumUp Ltd.",
        },
        "varRegexps": Object {
          "NAME": /SumUp Ltd\\\\\\./,
        },
      },
    ],
    "react-hooks/exhaustive-deps": "warn",
    "react-hooks/rules-of-hooks": "error",
  },
}
`;
